examples/ddp2pipe/ddp2pipe.py

```diff
@@ -0,0 +1,299 @@
+import argparse
+import os
+import socket
+
+import torch
+import torch.distributed
+import torch.distributed.rpc as rpc
+import torch.multiprocessing as mp
+import torch.nn.functional as F
+from torch import nn, optim
+from torch.nn.parallel import DistributedDataParallel as DDP
+from torch.utils.data import DistributedSampler
+from torchvision import datasets
+from torchvision.transforms import transforms
+from tqdm import tqdm
+
+from pippy import Pipe, PipelineDriverFillDrain, annotate_split_points, PipeSplitWrapper
+from pippy.microbatch import TensorChunkSpec, CustomReducer
+from pippy.utils import tp_transports
+
+USE_TQDM = True
+
+DIMS = [28 * 28, 500, 250, 100, 50, 25, 10]
+
+
+class DDPModel(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.mod = nn.Sequential(
+            *[layer for i in range(2) for layer in [nn.Linear(DIMS[i], DIMS[i + 1], bias=True), nn.ReLU()]]
+        )
+
+    def forward(self, x):
+        return self.mod(x)
+
+
+class PipeModel(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.mod = nn.Sequential(
+            *[layer for i in range(2, len(DIMS) - 1) for layer in
+              [nn.Sequential(nn.Linear(DIMS[i], DIMS[i + 1], bias=True), nn.ReLU())]]
+        )
+
+    def forward(self, x, labels):
+        logits = self.mod(x)
+        loss = F.cross_entropy(logits, labels)
+        return logits, loss
+
+
+class DDP2PipeConnector(nn.Module):
+    def __init__(self, pp_ranks, pp_group, dp_group_size):
+        super().__init__()
+        self.pp_ranks = pp_ranks
+        self.pp_group = pp_group
+
+        rank = torch.distributed.get_rank()
+        driver_rank = self.pp_ranks[0]
+
+        if rank == driver_rank:
+            self.pipe_model = PipeModel()
+
+            for i in range(1, 4):
+                annotate_split_points(self.pipe_model, {f'mod.{i}': PipeSplitWrapper.SplitPoint.BEGINNING})
+
+            # print(self.pipe_model)
+
+            self.pipe = Pipe.from_tracing(self.pipe_model, output_loss_value_spec=(False, True))
+            args_chunk_spec = (TensorChunkSpec(0), TensorChunkSpec(0))
```

Comment:
Oh this is very similar to @wanchaol 's Placement, maybe we can merge eventually?

Reply:
Yeah, agreed. looks like this is user facing API, it would be nice to merge these two concepts so that they have same API to the user https://github.com/pytorch/tau/blob/main/spmd/tensor/placement_types.py#L29